File: provider/src/main/java/com/raygun/raygun4android/IRaygunMessageBuilder.java
package com.raygun.raygun4android;

import android.content.Context;

import com.raygun.raygun4android.messages.crashreporting.RaygunMessage;

import java.util.List;
import java.util.Map;

public interface IRaygunMessageBuilder {
    RaygunMessage build();

    IRaygunMessageBuilder setMachineName(String machineName);

    IRaygunMessageBuilder setExceptionDetails(Throwable throwable);

    IRaygunMessageBuilder setClientDetails();

    IRaygunMessageBuilder setEnvironmentDetails(Context context);

    IRaygunMessageBuilder setVersion(String version);

    IRaygunMessageBuilder setTags(List tags);

    IRaygunMessageBuilder setUserCustomData(Map userCustomData);

    IRaygunMessageBuilder setAppContext(String identifier);

    IRaygunMessageBuilder setUserInfo();

    IRaygunMessageBuilder setNetworkInfo(Context context);

    IRaygunMessageBuilder setGroupingKey(String groupingKey);
}

File: provider/src/main/java/com/raygun/raygun4android/IRaygunMessageBuilder.kt
package com.raygun.raygun4android

import android.content.Context
import com.raygun.raygun4android.messages.crashreporting.RaygunBreadcrumbMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunMessage

interface IRaygunMessageBuilder {
    fun build(): RaygunMessage?

    fun setMachineName(machineName: String?): IRaygunMessageBuilder?

    fun setExceptionDetails(throwable: Throwable?): IRaygunMessageBuilder?

    fun setClientDetails(): IRaygunMessageBuilder?

    suspend fun setEnvironmentDetails(context: Context): IRaygunMessageBuilder?

    fun setVersion(version: String?): IRaygunMessageBuilder?

    fun setTags(tags: List<*>?): IRaygunMessageBuilder?

    fun setCustomData(customData: Map<*, *>?): IRaygunMessageBuilder?

    fun setAppContext(identifier: String?): IRaygunMessageBuilder?

    fun setUserInfo(): IRaygunMessageBuilder?

    fun setNetworkInfo(context: Context): IRaygunMessageBuilder?

    fun setGroupingKey(groupingKey: String?): IRaygunMessageBuilder?

    fun setBreadcrumbs(breadcrumbs: List<RaygunBreadcrumbMessage>?): IRaygunMessageBuilder?
}
